struct_ops: add structOpsMeta to carry BTF hints for StructOpsMap creation

shun159 opened the pull request:
1. Verify creating a StructOpsMap from a hand-crafted MapSpec with metadata.
2. Resolve attr.BtfVmlinuxValueTypeId from vmlinux BTF before MapCreate.
3. Value population is deferred to follow-up PRs.

see: https://github.com/cilium/ebpf/discussions/1502

shun159 commented:
> This PR is missing the part where program fds are inserted into the struct ops map. Without that it's hard to tell what is going on.

@lmb yeah, agreed.
I was trying to keep the PR small, but I see how it’s unclear without kern_vdata population.
then, let's proceed with including the changes so that make it clear.

shun159 commented on collection.go:
```diff
@@ -303,6 +308,11 @@ func (cs *CollectionSpec) LoadAndAssign(to interface{}, opts *CollectionOptions)
 	}
 
+	// load programs to struct_ops maps
+	if err := loader.stOps.postLoad(loader.maps); err != nil {
```
let's make it be a part of populateDeferredMaps

shun159 commented on map.go:
```diff
@@ -556,6 +556,47 @@ func (spec *MapSpec) createMap(inner *sys.FD) (_ *Map, err error) {
+			return nil, fmt.Errorf("open vmlinux BTF: %w", err)
+		}
+
+		typeSpec, err := s.AnyTypeByName(meta.kernTypeName)
```
We should check if it is a `Struct`. fixed.

shun159 commented on collection_test.go:
```diff
@@ -771,3 +771,40 @@ func ExampleCollectionSpec_LoadAndAssign() {
+				MaxEntries: 1,
+				Contents: []MapKV{
+					{Key: uint32(0), Value: structOpsMeta{
+						userTypeName: "bpf_dummy_ops",
```
For now, I'm proceeding as above. Since dummy_ops doesn't allow linking and programs need an AttachType specified before loading, I'm first testing whether a map can be created from the MapSpec. At a later step, I'll switch the tests to use testmod_ops.

lmb replied:
Follow up question: what does the value BTF look like for a struct ops map defined in C? I found https://github.com/torvalds/linux/blob/0cc53520e68bea7fb80fdc6bdf8d226d1b6a98d9/tools/testing/selftests/sched_ext/maybe_null.bpf.c is that the common use case?

I think that I misunderstood how struct ops maps are created: they don't really appear in the source code, instead we need the magic .struct_ops.link sections?

shun159 commented on helpers_test.go:
```diff
@@ -44,6 +44,36 @@ func requireTestmod(tb testing.TB) {
 	}
 }
 
+var haveStructOpsDummy = sync.OnceValues(func() (bool, error) {
```
@lmb this is not needed. removed for now.

shun159 commented:
vimto tests on 6.1 and 6.6 has passed on my setup, so I feel this is a temporary error.

shun159 commented:
> @shun159 please ping me explicitly when you need another review.

Hi @lmb I believe I've addressed most of your feedbacks earlier. 
However, after switching to use AttachTo, it looks like elf_reader_test is now failing. 
My assumption is that if the AttachTo string doesn't match the expected format introduced by this PR, we could simply ignore it, in which case the test itself should pass but I would appreciate your thought how we should proceed with this?

https://github.com/cilium/ebpf/actions/runs/17712680837/job/50333512472?pr=1845

lmb commented on collection.go:
```diff
@@ -697,6 +741,165 @@ func (cl *collectionLoader) populateDeferredMaps() error {
+
+		kernMember, ok := kernIndexByName[m.Name]
+		if !ok {
+			continue
```
Is this what libbpf does as well? Seems a bit prone to breakage.

lmb commented on collection.go:
```diff
@@ -697,6 +741,165 @@ func (cl *collectionLoader) populateDeferredMaps() error {
+			continue
+		}
+
+		if kernMember.BitfieldSize > 0 {
```
Move this to where kernIndexByName is built?

lmb commented on collection.go:
```diff
@@ -697,6 +741,165 @@ func (cl *collectionLoader) populateDeferredMaps() error {
+			return nil, fmt.Errorf("failed to resolve size of %s: %w", kernMember.Name, err)
+		}
+
+		if sz != kernSz {
```
This check is pretty basic. Is that all that libbpf does as well?

Might make sense to call `btf.CheckTypeCompatibility` if libbpf does something similar.

lmb commented on collection.go:
```diff
@@ -697,6 +741,165 @@ func (cl *collectionLoader) populateDeferredMaps() error {
+			return nil, fmt.Errorf("size mismatch for %s: %d != %d", m.Name, sz, kernSz)
+		}
+
+		src := int(m.Offset / 8)
```
.Bytes()

lmb reviewed (commented):
Looks pretty good! There are a couple of comments which you didn't address. Please go through the conversations and resolve them if you've actioned them.

To fix the CI breakage: for now it's ok to set `AttachTo = ""` if were dealing with a struct ops program in TestLibBPFCompat.

lmb reviewed (changes requested):
Looks pretty good! There are a couple of comments which you didn't address. Please go through the conversations and resolve them if you've actioned them.

To fix the CI breakage: for now it's ok to set `AttachTo = ""` if were dealing with a struct ops program in TestLibBPFCompat.

lmb commented:
Please rebase on top of https://github.com/cilium/ebpf/pull/1865. The end result will be that you shouldn't have to call `btf.LoadKernelSpec` explicitly anymore.

shun159 commented:
> Please rebase on top of #1865. The end result will be that you shouldn't have to call `btf.LoadKernelSpec` explicitly anymore.

done

shun159 commented:
@lmb Since the comment disappeared, I’ll respond by quoting below.

```
> +		return nil, fmt.Errorf("member %s not found in %s", innerName, to.Name)
+	}
+
+	kernIndexByName := make(map[string]btf.Member, len(inner.Members))
+	for _, m := range inner.Members {
+		kernIndexByName[m.Name] = m
+	}
+
+	for _, m := range from.Members {
+		if m.BitfieldSize > 0 {
+			return nil, fmt.Errorf("bitfield %s not supported", m.Name)
+		}
+
+		kernMember, ok := kernIndexByName[m.Name]
+		if !ok {
+			continue
Is this what libbpf does as well? Seems a bit prone to breakage.
```

From my understanding, libbpf seems to handle it as follows:

1. if a field in the “user struct” is not found in the kernel struct and its value is non-zero, it results in an error.
2. explicitly checks for matching type kinds.

shun159 commented:
At the moment, we only have a test that writes progFD into kern_vdata, but please let me also add some test to check whether scalar values can be written.

shun159 commented:
> @lmb Since the comment disappeared, I’ll respond by quoting below.
> 
> ```
> > +		return nil, fmt.Errorf("member %s not found in %s", innerName, to.Name)
> +	}
> +
> +	kernIndexByName := make(map[string]btf.Member, len(inner.Members))
> +	for _, m := range inner.Members {
> +		kernIndexByName[m.Name] = m
> +	}
> +
> +	for _, m := range from.Members {
> +		if m.BitfieldSize > 0 {
> +			return nil, fmt.Errorf("bitfield %s not supported", m.Name)
> +		}
> +
> +		kernMember, ok := kernIndexByName[m.Name]
> +		if !ok {
> +			continue
> Is this what libbpf does as well? Seems a bit prone to breakage.
> ```
> 
> ```
> > +
> +		if kernMember.BitfieldSize > 0 {
> +			return nil, fmt.Errorf("bitfield %s not supported in kern struct", kernMember.Name)
> +		}
> +
> +		sz, err := btf.Sizeof(m.Type)
> +		if err != nil {
> +			return nil, fmt.Errorf("failed to resolve size of %s: %w", m.Name, err)
> +		}
> +
> +		kernSz, err := btf.Sizeof(kernMember.Type)
> +		if err != nil {
> +			return nil, fmt.Errorf("failed to resolve size of %s: %w", kernMember.Name, err)
> +		}
> +
> +		if sz != kernSz {
> This check is pretty basic. Is that all that libbpf does as well?
> 
> Might make sense to call btf.CheckTypeCompatibility if libbpf does something similar.
> ```
> 
> From my understanding, libbpf seems to handle it as follows:
> 
> 1. if a field in the “user struct” is not found in the kernel struct and its value is non-zero, it results in an error.
> 2. explicitly checks for matching type kinds.
> 
> It may be need to keep the parsed result of the ELF somewhere.

@lmb 
From the VAR section, we can obtain information about each field of the “user struct” as shown below. By passing such section to CollectionSpec or similar, it seems possible to achieve the same functionality. I’d appreciate your thoughts on this.

```
[10] VAR 'null_sched_init.____fmt' type_id=9, linkage=static
[11] STRUCT 'sched_ext_ops' size=424 vlen=39
        'select_cpu' type_id=12 bits_offset=0
        'enqueue' type_id=21 bits_offset=64
        'dequeue' type_id=21 bits_offset=128
        'dispatch' type_id=23 bits_offset=192
        'tick' type_id=25 bits_offset=256
        'runnable' type_id=21 bits_offset=320
        'running' type_id=25 bits_offset=384
        'stopping' type_id=27 bits_offset=448
        'quiescent' type_id=21 bits_offset=512
        'yield' type_id=29 bits_offset=576
        'core_sched_before' type_id=29 bits_offset=640
        'set_weight' type_id=31 bits_offset=704
        'set_cpumask' type_id=36 bits_offset=768
        'update_idle' type_id=41 bits_offset=832
        'cpu_acquire' type_id=43 bits_offset=896
        'cpu_release' type_id=47 bits_offset=960
        'init_task' type_id=51 bits_offset=1024
        'exit_task' type_id=55 bits_offset=1088
        'enable' type_id=25 bits_offset=1152
        'disable' type_id=25 bits_offset=1216
        'dump' type_id=59 bits_offset=1280
        'dump_cpu' type_id=63 bits_offset=1344
        'dump_task' type_id=65 bits_offset=1408
        'cgroup_init' type_id=67 bits_offset=1472
        'cgroup_exit' type_id=73 bits_offset=1536
        'cgroup_prep_move' type_id=75 bits_offset=1600
        'cgroup_move' type_id=77 bits_offset=1664
        'cgroup_cancel_move' type_id=77 bits_offset=1728
        'cgroup_set_weight' type_id=79 bits_offset=1792
        'cpu_online' type_id=81 bits_offset=1856
        'cpu_offline' type_id=81 bits_offset=1920
        'init' type_id=83 bits_offset=1984
        'exit' type_id=85 bits_offset=2048
        'dispatch_max_batch' type_id=33 bits_offset=2112
        'flags' type_id=18 bits_offset=2176
        'timeout_ms' type_id=33 bits_offset=2240
        'exit_dump_len' type_id=33 bits_offset=2272
        'hotplug_seq' type_id=18 bits_offset=2304
        'name' type_id=89 bits_offset=2368
```

shun159 commented:
@lmb iPlease review it when you have time.

shun159 commented on map.go:
```diff
@@ -560,6 +560,41 @@ func (spec *MapSpec) createMap(inner *sys.FD) (_ *Map, err error) {
+				attr.ValueTypeBtfObjFd = int32(module.FD())
+			}
+
+			// TODO: This is required to make the map load, but doesn't make
```
I've confirmed the code comment added by @lmb .

lmb commented on prog.go:
```diff
@@ -381,15 +383,51 @@ func newProgramWithOptions(spec *ProgramSpec, opts ProgramOptions, c *btf.Cache)
+			if targetMember == "" {
+				return nil, fmt.Errorf("struct_ops: AttachTo must be '<ops>:<member>' (got %s)", spec.AttachTo)
+			}
+		}
```
@shun159 please update the ELF reader PR based on this.

shun159 replied:
@lmb ack.

dylandreimerink commented on collection.go:
```diff
@@ -697,6 +707,208 @@ func (cl *collectionLoader) populateDeferredMaps() error {
+		return err
+	}
+	mapSpec.Contents = nil
+	runtime.KeepAlive(cl.programs)
```
> I think for now we should force BPF_F_LINK by only allowing the ELF reader to consume .link sections. Tests should also only do F_LINK. I don't see how to make the non-link semantics work, they are just too broken / special. Maybe people need to "manually" create such maps if they need further backwards compat.

Yes, agreed. Lets throw an error if we do not see `BPF_F_LINK` when loading the map. Worst case people will need the add the flag to the spec manually in Go.

If we do go that route, we are missing the logic to create the link. We can add that in a followup PR. So this PR adds the logic to load a spec. https://github.com/cilium/ebpf/pull/1869 has the logic to go from ELF to spec. And some future PR will add the link logic to attach the loaded program/map. Sound good?
